### ISSUE TITLE
Why the size of Subgroups on the host is 1?

### ISSUE THREAD

**Issue by MajidSalimi** [AdaptiveCpp/AdaptiveCpp#592]
As the equivalence of CUDA cooperative groups, SYCL subgroups can be used on device and host. But the problem is that on the host side, the size of the subgroup is always "1". What is the reason for this? Is there any solution to consider host and device the same while using subgroups and resolve the problem?

**Comment by illuhad** [comment]
The subgroup size in SYCL is a property of the combination of the SYCL implementation, backend, and hardware. Subgroups should not be thought of as something that can be set or controlled by the user (apart from the fact that the user can query subgroup size and then potentially invoke different kernels), as in general the subgroup size depends e.g. on the hardware SIMD size. Implementations are free to choose the subgroup sizes that they support based on backend/hardware.

The subgroup size on CPU is 1 because hipSYCL's CPU backend relies on OpenMP, and therefore on an external OpenMP compiler to do the actual code generation. This has portability advantages.
hipSYCL therefore does not control vectorization on CPU. Different parts of a kernel in fact may even be vectorized differently.
Exposing a subgroup size greater than 1 might also require implementing forward progress guarantees for work items within a subgroup that cannot be done easily.

Note that even if the subgroup size is reported as 1, it does not necessarily mean that your code is not vectorized. However, note that the `nd_range` parallel for model is not performance portable, and should be avoided - especially if barriers are involved.

In summary, I don't see a subgroup size of 1 on CPU as a problem - it's just a characteristic of the backend, just like a fixed subgroup size of 32 is a characteristic of NVIDIA hardware. If your code makes assumptions about fixed subgroup sizes, you are already  deep in the business of creating potentially unportable device-specific optimizations and code paths :)
